File: app/src/main/java/com/pr0gramm/app/ui/ContactActivity.java
package com.pr0gramm.app.ui;

import android.content.res.ColorStateList;
import android.os.Bundle;
import android.support.v4.content.ContextCompat;
import android.support.v4.view.ViewCompat;
import android.support.v7.app.ActionBar;
import android.util.Patterns;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.RadioButton;
import android.widget.TextView;

import com.google.common.base.Optional;
import com.pr0gramm.app.ActivityComponent;
import com.pr0gramm.app.BuildConfig;
import com.pr0gramm.app.R;
import com.pr0gramm.app.feed.Nothing;
import com.pr0gramm.app.services.ContactService;
import com.pr0gramm.app.services.FeedbackService;
import com.pr0gramm.app.services.ThemeHelper;
import com.pr0gramm.app.services.UserService;
import com.pr0gramm.app.ui.base.BaseAppCompatActivity;

import java.util.Collections;
import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.BindViews;
import butterknife.OnClick;
import rx.Observable;
import rx.functions.Actions;

import static com.pr0gramm.app.services.ThemeHelper.theme;
import static com.pr0gramm.app.ui.dialogs.ErrorDialogFragment.defaultOnError;
import static com.pr0gramm.app.ui.fragments.BusyDialogFragment.busyDialog;

/**
 */
public class ContactActivity extends BaseAppCompatActivity {
    @Inject
    FeedbackService feedbackService;

    @Inject
    UserService userService;

    @Inject
    ContactService contactService;

    @BindView(R.id.submit)
    Button buttonSubmit;

    @BindView(R.id.feedback_name)
    EditText vName;

    @BindView(R.id.feedback_text)
    EditText vText;

    @BindView(R.id.feedback_email)
    EditText vMail;

    @BindView(R.id.feedback_subject)
    EditText vSubject;

    @BindView(R.id.action_feedback_general)
    RadioButton generalFeedbackRadioButton;

    @BindViews({R.id.feedback_email, R.id.feedback_name, R.id.feedback_subject, R.id.feedback_text})
    List<TextView> groupAllTextViews;

    @BindViews({R.id.feedback_email, R.id.feedback_name, R.id.feedback_subject, R.id.feedback_deletion_hint})
    List<View> groupAll;

    @BindViews({R.id.feedback_email, R.id.feedback_subject, R.id.feedback_deletion_hint})
    List<View> groupNormalSupport;

    @BindViews({R.id.feedback_name})
    List<View> groupAppNotLoggedIn;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        setTheme(theme().basic);
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_feedback);

        ActionBar actionbar = getSupportActionBar();
        if (actionbar != null) {
            actionbar.setDisplayHomeAsUpEnabled(true);
        }

        int primary = ContextCompat.getColor(this, ThemeHelper.accentColor());
        ViewCompat.setBackgroundTintList(buttonSubmit, ColorStateList.valueOf(primary));

        // register all the change listeners
        for (TextView textView : groupAllTextViews) {
            textView.addTextChangedListener(new SimpleTextWatcher() {
                @Override
                public void onTextChanged(CharSequence s, int start, int before, int count) {
                    updateSubmitButtonActivation();
                }
            });
        }

        Optional<String> name = userService.getName();
        if (name.isPresent()) {
            vName.setText(name.get());
        }

        applyViewVisibility();
    }

    @OnClick({R.id.action_feedback_general, R.id.action_feedback_app})
    void applyViewVisibility() {
        List<View> activeViews;
        if (isNormalSupport()) {
            activeViews = groupNormalSupport;
        } else if (userService.isAuthorized()) {
            activeViews = Collections.emptyList();
        } else {
            activeViews = groupAppNotLoggedIn;
        }

        for (View view : groupAll) {
            view.setVisibility(activeViews.contains(view) ? View.VISIBLE : View.GONE);
        }

        updateSubmitButtonActivation();
    }

    void updateSubmitButtonActivation() {
        boolean enabled = true;
        for (TextView view : groupAllTextViews) {
            if (view.getVisibility() == View.VISIBLE) {
                if (view.getText().toString().trim().isEmpty()) {
                    enabled = false;
                }
            }
        }

        if (vMail.getVisibility() == View.VISIBLE) {
            if (!Patterns.EMAIL_ADDRESS.matcher(vMail.getText()).matches())
                enabled = false;
        }

        buttonSubmit.setEnabled(enabled);
    }

    private boolean isNormalSupport() {
        return generalFeedbackRadioButton.isChecked();
    }

    @Override
    protected void injectComponent(ActivityComponent appComponent) {
        appComponent.inject(this);
    }

    @OnClick(R.id.submit)
    void submitClicked() {
        String feedback = vText.getText().toString().trim();

        Observable<Nothing> response;
        if (isNormalSupport()) {
            String email = vMail.getText().toString().trim();
            String subject = vSubject.getText().toString().trim();

            feedback += "\n\nGesendet mit der App v" + BuildConfig.VERSION_NAME;

            response = contactService.contactFeedback(email, subject, feedback).toObservable();
        } else {
            String name = userService.getName().or(vName.getText().toString().trim());
            response = feedbackService.post(name, feedback);
        }

        response.compose(bindToLifecycleAsync())
                .lift(busyDialog(this))
                .subscribe(Actions.empty(), defaultOnError(), this::onSubmitSuccess);
    }

    private void onSubmitSuccess() {
        DialogBuilder.start(this)
                .content(R.string.feedback_sent)
                .positive(R.string.okay, di -> finish())
                .onCancel(di -> finish())
                .show();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            finish();
            return true;
        }

        return super.onOptionsItemSelected(item);
    }
}

File: app/src/main/java/com/pr0gramm/app/ui/ContactActivity.kt
package com.pr0gramm.app.ui

import android.app.Activity
import android.content.Context
import android.os.Bundle
import android.util.Patterns
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.TextView
import androidx.core.view.isVisible
import com.pr0gramm.app.databinding.ActivityFeedbackBinding
import com.pr0gramm.app.services.ContactService
import com.pr0gramm.app.services.ThemeHelper
import com.pr0gramm.app.services.UserService
import com.pr0gramm.app.ui.base.BaseAppCompatActivity
import com.pr0gramm.app.ui.base.bindViews
import com.pr0gramm.app.ui.base.launchWhenStarted
import com.pr0gramm.app.ui.base.withViewDisabled
import com.pr0gramm.app.util.Linkify
import com.pr0gramm.app.util.di.instance
import com.pr0gramm.app.util.hideSoftKeyboard
import com.pr0gramm.app.util.matches
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.NonCancellable
import kotlinx.coroutines.withContext

/**
 */
class ContactActivity : BaseAppCompatActivity("ContactActivity") {
    private val contactService: ContactService by instance()
    private val userService: UserService by instance()

    private val views by bindViews(ActivityFeedbackBinding::inflate)

    private val groupAllInputViews: List<TextView>
        get() = listOf(views.feedbackEmail, views.feedbackName, views.feedbackSubject, views.feedbackText)

    private val groupAll: List<View>
        get() = listOf(views.feedbackEmail, views.feedbackName, views.feedbackSubject, views.feedbackDeletionHint)

    private val groupNormalLoggedIn: List<View>
        get() = listOf(views.feedbackSubject, views.feedbackDeletionHint)

    private val groupNormalLoggedOut: List<View>
        get() = listOf(views.feedbackEmail, views.feedbackSubject, views.feedbackDeletionHint)

    override fun onCreate(savedInstanceState: Bundle?) {
        setTheme(ThemeHelper.theme.basic)
        super.onCreate(savedInstanceState)

        setContentView(views)

        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        // val primary = ContextCompat.getColor(this, ThemeHelper.accentColor)
        // ViewCompat.setBackgroundTintList(views.submit, ColorStateList.valueOf(primary))

        // register all the change listeners
        for (textView in groupAllInputViews) {
            textView.addTextChangedListener(object : SimpleTextWatcher() {
                override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                    updateSubmitButtonActivation()
                }
            })
        }

        Linkify.linkify(views.feedbackDeletionHint)

        views.submit.setOnClickListener { submitClicked() }

        views.faqCategory.adapter = CategoriesAdapter(this)
        views.faqCategory.setSelection(0)

        views.faqCategory.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onNothingSelected(parent: AdapterView<*>?) = Unit
            override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
                updateSubmitButtonActivation()
            }
        }

        userService.name?.let { views.feedbackName.setText(it) }

        applyViewVisibility()
    }

    private class CategoriesAdapter(context: Context) : ArrayAdapter<Category>(
            context, android.R.layout.simple_spinner_dropdown_item, faqCategories) {

        override fun isEnabled(position: Int): Boolean = position > 0

        override fun getDropDownView(position: Int, convertView: View?, parent: ViewGroup): View {
            return super.getDropDownView(position, convertView, parent).also { view ->
                view.isEnabled = position > 0
            }
        }
    }

    private fun applyViewVisibility() {
        val activeViews: List<View> = when {
            userService.isAuthorized -> groupNormalLoggedIn
            else -> groupNormalLoggedOut
        }

        for (view in groupAll) {
            view.isVisible = view in activeViews
        }

        updateSubmitButtonActivation()
    }

    private fun updateSubmitButtonActivation() {
        // the button is enabled if there is no visible text view that has
        // an empty input.
        var enabled = groupAllInputViews.none { view ->
            view.isVisible && view.text.toString().isBlank()
        }

        if (views.feedbackEmail.isVisible && !views.feedbackEmail.text.matches(Patterns.EMAIL_ADDRESS)) {
            enabled = false
        }

        val faqCategory = views.faqCategory.selectedItem as Category
        if (faqCategory.category == "none") {
            enabled = false
        }

        views.submit.isEnabled = enabled
    }

    private fun submitClicked() {
        // hide keyboard when sending
        hideSoftKeyboard()

        launchWhenStarted(busyIndicator = true) {
            withViewDisabled(views.submit) {
                sendFeedback()
                onSubmitSuccess()
            }
        }
    }

    private suspend fun sendFeedback() {
        val category = views.faqCategory.selectedItem as Category
        val email = views.feedbackEmail.text.toString().trim()
        val feedback = views.feedbackText.text.toString().trim()
        var subject = views.feedbackSubject.text.toString().trim()

        if (category.category == "app") {
            subject = "[app] $subject"
        }

        withContext(Dispatchers.IO + NonCancellable) {
            contactService.post(email, subject, feedback)
        }
    }

    private fun onSubmitSuccess() {
        setResult(Activity.RESULT_OK)
        finish()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            finish()
            return true
        }

        return super.onOptionsItemSelected(item)
    }
}

private class Category(val category: String, val text: String) {
    override fun toString(): String = text
}

private val faqCategories = listOf(
    Category("none", "Kategorie auswählen"),
    Category("app", "App"),
    Category("pr0gramm", "pr0gramm"),
)
